resolwe/flow/managers/base.py

```diff
@@ -87,11 +198,133 @@ def _include_environment_variables(self, program):
         export_commands = ['export {}="{}"'.format(key, value.replace('"', '\"')) for key, value in env_vars.items()]
         return os.linesep.join(export_commands) + os.linesep + program
 
-    def run(self, data_id, script, priority='normal', run_sync=False, verbosity=1):
+    def run(self, data_id, dest_dir, argv, priority='normal', run_sync=False, verbosity=1):
         """Run process."""
         raise NotImplementedError('`run` function not implemented')
 
-    def communicate(self, run_sync=False, verbosity=1):
+    def _get_per_data_dir(self, dir_base, data_id):
+        """Return the given base directory extended with a per-data component."""
+        result = getattr(settings, 'FLOW_EXECUTOR', {}).get(dir_base, '')
+        return os.path.join(result, str(data_id))
+
+    def _prepare_data_dir(self, data_id):
+        """Prepare destination directory where the data will live."""
+        output_path = self._get_per_data_dir('DATA_DIR', data_id)
+        dir_mode = getattr(settings, 'FLOW_EXECUTOR', {}).get('DATA_DIR_MODE', 0o755)
+        os.mkdir(output_path, mode=dir_mode)
+        # os.mkdir is not guaranteed to set the given mode
+        os.chmod(output_path, dir_mode)
+        return output_path
+
+    def _prepare_context(self, data_id, data_dir, runtime_dir, **kwargs):
+        """Prepare settings and constants jsons for the executor."""
+        settings_dict = dict(kwargs)
+        settings_dict.update(self._marshal_settings())
+        settings_dict['DATA_DIR'] = data_dir
+
+        # add scheduling classes
+        settings_dict['Process_sched_class'] = {
+            k: getattr(Process, k) for k in dir(Process)
+            if k.startswith('SCHEDULING_CLASS_') and isinstance(getattr(Process, k), str)
+        }
+        # add Data status constants
+        settings_dict['Data_status'] = {
+            k: getattr(Data, k) for k in dir(Data)
+            if k.startswith('STATUS_') and isinstance(getattr(Data, k), str)
+        }
+        # required plumbing settings
+        settings_dict['REDIS_CHANNEL_PAIR'] = state.MANAGER_EXECUTOR_CHANNELS
+
+        # extend the settings with whatever the executor wants
+        settings_dict.update(self.executor.make_settings(data_id))
+
+        # save the settings into a json in the data dir
+        json.dump(settings_dict, open(os.path.join(runtime_dir, 'settings.json'), 'wt'), cls=SettingsJSONifier)
+
+    def _prepare_executor(self, data_id, executor):
+        """Copy executor sources into the destination directory."""
+        import resolwe.flow.executors as executor_module
+
+        exec_dir = executor_module.__path__[0]
```

Comment:
An empty `__path__` attribute makes no sense since the package is not runtime constructed; it must always be loaded from a directory and thus always has a usable `__path__` attribute. If `__path__` is empty, then the manager as a whole is in an undefined state and can not proceed.

Reply:
Thanks for your explanation. Where do we handle "manager as a whole in an undefined state"?

BTW, I've got this error when trying to run a test:
```
Traceback (most recent call last):
  File "/home/tadej/.virtualenvs/resolwe/lib/python3.5/site-packages/channels/worker.py", line 119, in run
    consumer(message, **kwargs)
  File "/home/tadej/Genialis/resolwe/resolwe/flow/managers/base.py", line 162, in __init__
    super(BaseManager, self).__init__(*args, **kwargs)
  File "/home/tadej/.virtualenvs/resolwe/lib/python3.5/site-packages/channels/generic/base.py", line 32, in __init__
    self.dispatch(message, **kwargs)
  File "/home/tadej/.virtualenvs/resolwe/lib/python3.5/site-packages/channels/generic/base.py", line 70, in dispatch
    return self.get_handler(message, **kwargs)(message, **kwargs)
  File "/home/tadej/Genialis/resolwe/resolwe/flow/managers/base.py", line 285, in handle_control_event
    self._communicate(**message.content[WorkerProtocol.COMMUNICATE_EXTRA])
  File "/home/tadej/Genialis/resolwe/resolwe/flow/managers/base.py", line 397, in _communicate
    executor_module, runtime_dir = self._prepare_executor(data.id, executor)
  File "/home/tadej/Genialis/resolwe/resolwe/flow/managers/base.py", line 251, in _prepare_executor
    exec_dir = executor_module.__path__[0]
TypeError: '_NamespacePath' object does not support indexing
```

Why has this happened?